File: src/screens/title.lua
-- Skelly title/loading screen.
--
-- By Chris Herborth (https://github.com/Taffer)
-- MIT license, see LICENSE.md for details.

local Class = require 'lib/middleclass/middleclass'
local ScreenBase = require 'src/screens/base'

local rsrc_list = require 'src/rsrc_list'

local function loader(resource, file_list)
    local yield = coroutine.yield

    -- Load the files listed into the resource table.
    for k,v in pairs(file_list.images) do
        resource.images[k] = love.graphics.newImage(v)
        yield(v)
    end

    for k,v in pairs(file_list.music) do
        resource.music[k] = love.audio.newSource(v, 'stream')
        yield(v)
    end

    for k,v in pairs(file_list.sounds) do
        resource.sounds[k] = love.audio.newSource(v, 'static')
        yield(v)
    end

    return resource.text.title.loading_done
end

local TitleScreen = Class('TitleScreen', ScreenBase)

function TitleScreen:initialize(resources)
    ScreenBase.initialize(self, resources)

    self.resources.fonts.skelly_title = love.graphics.newFont('graphics/Gypsy Curse.ttf', 144)
    self.resources.images.skelly_title = love.graphics.newImage('graphics/Gersdorff_Feldbuch_skeleton.png')

    self.loading_x = 16
    self.loading_y = love.graphics.getHeight() - 16 - self.resources.fonts.default_mono:getHeight()

    self.skelly_text = self.resources.text.skelly_title
    self.subtitle_text = self.resources.text.title.subtitle_text
    self.loading_text = self.resources.text.title.loading_text

    self.alpha = 0 -- Alpha level for the fade-in/out animation.
    self.ticks = 0
    self.pi_over_180 = math.pi / 180
    self.degrees_per_second = 45

    self.loaded_resource = ""
    self.loading_finished = false
    self.loading_routine = nil
end

-- Render this screen's contents.
function TitleScreen:draw()
    -- Premature optimization:
    local rsrc = self.resources
    local font_mono = rsrc.fonts.default_mono
    local font_title = rsrc.fonts.skelly_title
    local image_title = rsrc.images.skelly_title

    love.graphics.clear(0, 0, 0, 1)

    love.graphics.setColor(1, 1, 1, self.alpha)
    love.graphics.draw(image_title, 0, 0)

    local screen_width = love.graphics.getWidth()
    local width = font_mono:getWidth(self.loading_text)
    local x = (screen_width - width) / 2
    love.graphics.setFont(font_mono)
    love.graphics.print(self.loading_text .. ' ' .. (self.loaded_resource or ""),
        self.loading_x, self.loading_y)

    width = font_mono:getWidth(self.subtitle_text)
    x = (screen_width - width) / 2
    love.graphics.print(self.subtitle_text, x, 200)

    width = font_title:getWidth(self.skelly_text)
    x = (screen_width - width) / 2
    love.graphics.setFont(font_title)
    love.graphics.print(self.skelly_text, x, 40)
end

-- Update the screen.
function TitleScreen:update(dt)
    self.ticks = self.ticks + dt

    local degrees = self.ticks * self.degrees_per_second -- 1 second = 90 degrees

    if degrees >= 90 and not self.loading_finished then
        -- Pause the animation until loading is done.
        degrees = 90
    end

    self.alpha = math.sin(degrees * self.pi_over_180)

    if degrees > 180 then -- sin(180 degrees) is back to 0 alpha
        self.exit_screen = true
    end

    -- Load resources.
    if self.loading_routine then
        local resume = coroutine.resume
        alive, self.loaded_resource = resume(self.loading_routine,self.resources, rsrc_list)
        if not alive then
            self.loading_finished = true
            self.loaded_resource = self.resources.text.title.loading_done

        end
    else
        self.loading_routine = coroutine.create(loader)
    end
end

-- Exit this screen?
function TitleScreen:exit()
    return self.exit_screen
end

-- Handle events.
--
-- If you handled it, return true; false means the event continues on to the
-- next handler.
function TitleScreen:handle(event)
    if event == 'escape' then
        -- Escape doesn't kick you out until loading is done, sorry.
        if self.loading_finished then
            self.exit_screen = true
            return true
        end
    end

    return ScreenBase.handle(self, event)
end

return TitleScreen

File: src/screens/Title.py
''' Skelly title screen class.

By Chris Herborth (https://github.com/Taffer)
MIT license, see LICENSE.md for details.
'''

import pygame

from . import Base
from .. import RESOURCE_LIST
from ..ui import ColorFade
from ..ui import Label

BLACK = pygame.Color('black')
BLACK_ALPHA = pygame.Color(BLACK.r, BLACK.g, BLACK.g, 0)  # BLACK, but fully transparent
WHITE = pygame.Color('white')


def loader(game: any, file_list: dict, done_text: str) -> str:
    ''' Load the files in the RESOURCE_LIST.
    '''
    yield ''

    resource = game.resources

    for k, v in file_list['font_paths'].items():
        try:
            game.manager.add_font_paths(k, v['regular'], v['bold'], v['italic'], v['bold_italic'])
        except Exception as ex:
            print('Unable to add font paths: {0} {1}'.format(k, ex))
        yield k

    for k, v in file_list['fonts'].items():
        try:
            game.manager.preload_fonts([v])
        except Exception as ex:
            print('Unable to load font: {0} {1}'.format(k, ex))
        yield v['name']

    for k, v in file_list['images'].items():
        try:
            resource['images'][k] = pygame.image.load(v).convert_alpha()
        except Exception as ex:
            print('Unable to load image: {0} {1}'.format(k, ex))
        yield v

    for k, v in file_list['music'].items():
        try:
            resource['music'][k] = pygame.mixer.Sound(v)
        except Exception as ex:
            print('Unable to load music: {0} {1}'.format(k, ex))
        yield v

    for k, v in file_list['sounds'].items():
        try:
            resource['sounds'][k] = pygame.mixer.Sound(v)
        except Exception as ex:
            print('Unable to load sound: {0} {1}'.format(k, ex))
        yield v

    for k, v in file_list['maps'].items():
        resource['maps'][k] = v  # Map class loads from paths.
        yield v

    yield done_text


class Title(Base):
    def __init__(self, game: any) -> None:
        super().__init__(game)

        self.next_screen = 'Journey'

        self.game.resources['fonts']['skelly_title'] = pygame.freetype.Font('graphics/Gypsy Curse.ttf', 144)
        self.game.resources['images']['skelly_title'] = pygame.image.load('graphics/Gersdorff_Feldbuch_skeleton.png').convert()

        self.loading_x = 16
        self.loading_y = self.game.screen_height - 16 - self.game.resources['fonts']['default_mono'].get_sized_height()

        title_text = self.game.text.get_text('title')
        self.loading_text = title_text['loading_text']
        self.done_text = title_text['loading_done']

        self.fade = ColorFade(BLACK, BLACK_ALPHA, 1)  # 1 second fade
        self.fade_out = False

        self.loaded_resource = ""
        self.loading_finished = False
        self.loading_routine = loader(self.game, RESOURCE_LIST, self.done_text)

        self.add_title()

        self.ui = []
        self.loading_label = Label(self.loading_x, self.loading_y, self.loading_text, self.game.resources['fonts']['default_mono'],
                                   WHITE, 'left')
        self.ui.append(self.loading_label)

    def draw(self) -> None:
        self.game.surface.fill(BLACK)
        self.draw_title()

        for item in self.ui:
            item.draw()

        self.fade.draw()

    def update(self, dt: float) -> None:
        self.fade.update(dt)

        if self.fade_out:
            # If we're fading out...
            if self.fade.is_done():
                self.can_exit = True
        else:
            # If we're fading in...
            if self.fade.is_done():
                if self.loading_finished:
                    self.fade = ColorFade(BLACK_ALPHA, BLACK, 1)
                    self.fade_out = True

            if not self.loading_finished:
                try:
                    result = self.loading_routine.__next__()
                    self.loading_label.set_text('{0} {1}'.format(self.loading_text, result))
                except StopIteration:
                    self.loading_finished = True
